File: tests/conformance/language/control-flow.js
/*---
description: Tests conditional statements, loops, switch case, and break/continue.
---*/

// If-else
let a = 5;
let ifResult = "";
if (a > 10) {
    ifResult = "big";
} else if (a > 3) {
    ifResult = "medium";
} else {
    ifResult = "small";
}
assert.sameValue(ifResult, "medium", "if-else-if condition");

// Ternary
let ternary = a > 3 ? "big" : "small";
assert.sameValue(ternary, "big", "ternary operator");

// For loop
let sum = 0;
for (let i = 1; i <= 5; i++) {
    sum = sum + i;
}
assert.sameValue(sum, 15, "for loop sum");

// While loop
let w = 5;
while (w > 0) {
    w = w - 1;
}
assert.sameValue(w, 0, "while loop decrement");

// Do-while loop
let d = 0;
do {
    d = d + 1;
} while (d < 3);
assert.sameValue(d, 3, "do-while loop increment");

// For-in loop
let obj = { a: 1, b: 2 };
let keys = "";
for (let k in obj) {
    keys = keys + k;
}
assert(keys === "ab" || keys === "ba", "for-in loop over object keys");

// Switch statement
let day = 2;
let dayName = "";
switch (day) {
    case 1:
        dayName = "Mon";
        break;
    case 2:
        dayName = "Tue";
        break;
    default:
        dayName = "Other";
}
assert.sameValue(dayName, "Tue", "switch statement matching case 2");

// Switch default
let otherDay = 99;
let otherDayName = "";
switch (otherDay) {
    case 1:
        otherDayName = "Mon";
        break;
    default:
        otherDayName = "Other";
}
assert.sameValue(otherDayName, "Other", "switch statement matching default case");

// Break and continue
let bcSum = 0;
for (let i = 0; i < 10; i++) {
    if (i === 3) continue;
    if (i === 7) break;
    bcSum = bcSum + i;
}
assert.sameValue(bcSum, 18, "break and continue in loop");

File: tests/all_features/control_flow.rs
use tails::{TailsRuntime, Value};

#[test]
fn test_if_else() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let a = 5;
    if (a > 10) { "big"; } else if (a > 3) { "medium"; } else { "small"; }
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::string("medium"));
}

#[test]
fn test_ternary() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let a = 5;
    let ternary = a > 3 ? "big" : "small";
    ternary;
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::string("big"));
}

#[test]
fn test_for_loop() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let sum = 0;
    for (let i = 1; i <= 5; i++) { sum = sum + i; }
    sum;
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::Float(15.0));
}

#[test]
fn test_while_loop() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let w = 5;
    while (w > 0) { w = w - 1; }
    w;
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::Float(0.0));
}

#[test]
fn test_do_while_loop() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let d = 0;
    do { d = d + 1; } while (d < 3);
    d;
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::Float(3.0));
}

#[test]
fn test_for_in() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let obj = { a: 1, b: 2 };
    let keys = "";
    for (let k in obj) { keys = keys + k; }
    keys;
    "#,
    );
    assert!(r.is_ok());
    let val = r.unwrap();
    assert!(val == Value::string("ab") || val == Value::string("ba"));
}

#[test]
fn test_switch() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let day = 2;
    let dayName = "";
    switch (day) {
        case 1: dayName = "Mon"; break;
        case 2: dayName = "Tue"; break;
        default: dayName = "Other";
    }
    dayName;
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::string("Tue"));
}

#[test]
fn test_break_continue() {
    let mut rt = TailsRuntime::default();
    let r = rt.eval(
        r#"
    let sum = 0;
    for (let i = 0; i < 10; i++) {
        if (i === 3) continue;
        if (i === 7) break;
        sum = sum + i;
    }
    sum;
    "#,
    );
    assert!(r.is_ok());
    assert_eq!(r.unwrap(), Value::Float(18.0));
}
